fem/bilininteg.hpp

```diff
@@ -14,18 +14,81 @@
 
 #include "../config/config.hpp"
 #include "nonlininteg.hpp"
+#include "fespace.hpp"
+#include "kBilinIntegDiffusion.hpp"
 
 namespace mfem
 {
 
-/// Abstract base class BilinearFormIntegrator
-class BilinearFormIntegrator : public NonlinearFormIntegrator
+// *****************************************************************************
+// * Abstract Bilinear Form Integrator
+// *****************************************************************************
+class AbstractBilinearFormIntegrator : public NonlinearFormIntegrator
 {
-protected:
-   BilinearFormIntegrator(const IntegrationRule *ir = NULL) :
+public:
+   AbstractBilinearFormIntegrator(const IntegrationRule *ir = NULL) :
```

Comment:
Yes, I haven't understood really what `ir` to use, I'm still testing for `NULL` and if not construct it.

Reply:
Maybe @v-dobrev can clarify the use of this `IntegrationRule` and why it is acceptable to not set it?

Reply:
This is actually simple -- if the the internal `IntRule` (in the parent class `NonlinearFormIntegrator`) is NULL, then use the default quadrature rule. Otherwise use the provided integration rule, e.g. set with `SetIntRule()`.

Reply:
I guess it will clarify some other spots. By default `IntegrationRule`, do you mean:
```
const IntegrationRule& ir = intRules.Get(fes->GetFE(0)->GetGeomType(), order);
```
Because my problem was that at the point where I wanted to use `GetIntegrationRule()` I didn't have `fes`, nor `order`. Does that mean that every time we should check if there is a *special* `IntegrationRule` in the `Integrator` and if not carry with us the `fes` and `order` just in case there is not such an `IntegrationRule` provided? Wouldn't it be simpler to just force every `Integrator` to have an `IntegrationRule`? Whould that be a problem? Could the constructor construct the default `IntegrationRule` instead when `ir=NULL`? I find it very confusing that `GetIntegrationRule` returns `NULL` when it should return the default `IntegrationRule`.